### PR TITLE
ramips: mt7621: add support for ZyXEL WSM20

### PR DESCRIPTION
**Pull request by andyboeh**
The ZyXEL WSM20 aka Multy M1 is a cheap mesh router system by ZyXEL based on the MT7621 CPU.

Specifications
==============

SoC: MediaTek MT7621AT (880MHz)
RAM: 256MiB
Flash: 128MiB NAND
Wireless: 802.11ax (2x2 MT7915E DBDC)
Ethernet: 4x 10/100/1000 (MT7530)
Button: 1x WPS, 1x Reset, 1x LED On/Off
LED: 7 LEDs (3x white, 2x red, 2x green)

Installation
============

The device is cloud-managed, but there is a hidden local firmware upgrade
page in the OEM web interface. The device has to be registered in the
cloud in order to be able to access this page.

The system has a dual firmware design, there is no way to tell which
firmware is currently booted. Therefore, an -initramfs version is flashed
first.

1. Log into the OEM web GUI
2. Access the hidden upgrade page by navigating to
   https://192.168.212.1/gui/#/main/debug/firmwareupgrade
3. Upload the -initramfs-recovery.bin file and flash it
4. Wait for OpenWrt to boot and log in via SSH
5. Transfer the sysupgrade file via SCP
6. Run sysupgrade to install the image
7. Reboot and enjoy

Back to stock
=============

It is possible to flash back to stock, but a OEM firmware upgrade is
required. ZyXEL does not provide the link on its website, but the link
can be acquired from the OEM web GUI by analyzing the transferred JSON
objects.

It is then a matter of writing the firmware to Kernel2 and setting the
boot partition to FW2:

   mtd write zyxel.bin Kernel2
   echo -ne "\x02" | dd of=/dev/mtdblock7 count=1 bs=1 seek=4 conv=notrunc

Signed-off-by: Andreas Böhler <dev@aboehler.at>
Credits to forum users Annick and SirLouen for their initial work and testing on this device

### CODE REVIEW

**Review by SirLouen — APPROVED** [review]

**Comment by Djfe** [comment]
Awesome!
Somehow I missed this. I'm going to test this patch now on my device.

Did Zyxel send you their source code?
They promised to come back to me but I never received a second reply from them with the gpl source code so far :/

I learned a lot from reading your upgrade script :)
Could you explain `znet-header` maybe? Was this mostly found by reverse engineering and trial and error?
Or is the same header also used by other devices?

**Comment by SirLouen** [comment]
> Awesome! Somehow I missed this. I'm going to test this patch now on my device.
> 
> Did Zyxel send you their source code? They promised to come back to me but I never received a second reply from them with the gpl source code so far :/
> 
> I learned a lot from reading your upgrade script :) Could you explain `znet-header` maybe? Was this mostly found by reverse engineering and trial and error? Or is the same header also used by other devices?

Everything was done by reverse engineering and checking how it was done in other devices. This device showed so many similarities that was pretty straight forward

I also contacted ZyXEL back in the day but they never sent the DTS either. If you checked other firmwares you could see a clear pattern among all, I could boot it just by adjusting the kernel mem addresses through U-Boot on memoryu but could not make it load straight from boot on flash.

But suddenly Andy appeared and discovered two CRC that were the key for booting and copying this script from another device that used the same CRC mechanism (is in the same file). In fact other ZyXEL devices also had this CRC thing (but did the conversion in a separate script file not within the chipset mk file).

I think you should check this thread with all the info taken in the process
https://forum.openwrt.org/t/openwrt-for-zyxel-wsm20-multy-m1/

The last steps have been Andy trying to make it more "universal" installation (there are two variants, depending on where the official firmware was installed in two partitions). 

Now I don't understand what shall it take to make it through snapshot. It says it needs 6 reviews from people with write access. Isn't it too much? It doesn't even have 1 already. It feels it's going to take like 2 years.

**Comment by andyboeh** [comment]
> Now I don't understand what shall it take to make it through snapshot. It says it needs 6 reviews from people with write access. Isn't it too much? It doesn't even have 1 already. It feels it's going to take like 2 years.

That is on purpose: The review system on github is not used. A maintainer is going to add the commit to his staging tree,  then the staging tree is merged. 

> Did Zyxel send you their source code?

Nope, although they promised to send it in August or so, when it's ready.

> Could you explain `znet-header` maybe? Was this mostly found by reverse engineering and trial and error?
> Or is the same header also used by other devices?

Reversing only, but it's not very complicated. The header is unique, but there are similarities with other ZyXEL headers. The thread @SirLouen linked contains most of the description, including an easier-to-comprehend version in C that I made before porting it to shell code.

**Pull request merged**
